processing/src/main/java/io/druid/query/groupby/epinephelinae/ConcurrentGrouper.java

```diff
@@ -292,19 +325,413 @@ private boolean isParallelSortAvailable()
+              }
+            }
+          }
+          return false;
+        }
+        return true;
+      }
+
+      @Override
+      public Entry<KeyType> next()
+      {
+        return combinedIterator.next();
+      }
+    };
+  }
+
+  /**
+   * Find a minimum size of the buffer slice and corresponding combining degree and number of slices.  Note that each
+   * node in the combining tree is executed by different threads.  This method assumes that higher degree of parallelism
+   * can exploit better performance and find such a shape of the combining tree.
+   *
+   * @param combineBuffer                 entire buffer used for combining tree
+   * @param requiredMinimumBufferCapacity minimum buffer capacity for {@link StreamingMergeSortedGrouper}
+   * @param concurrencyHint               available degree of parallelism
+   * @param maxDegree                     maximum degree
+   *
+   * @return a pair of degree and number of buffers if found. Otherwise null.
+   */
+  @Nullable
+  private static Pair<Integer, Integer> findCombineDegreeAndNumBuffers(
+      ByteBuffer combineBuffer,
+      int requiredMinimumBufferCapacity,
+      int concurrencyHint,
+      int maxDegree
+  )
+  {
+    for (int degree = MINIMUM_COMBINE_DEGREE; degree <= maxDegree; degree++) {
+      // the number of available combine nodes is concurrencyHint because it's the max concurrency and we don't want to
+      // parallelize behind that.
+      final int requiredBufferNum = computeRequiredBufferNum(concurrencyHint, degree);
+      final int expectedSliceSize = combineBuffer.capacity() / requiredBufferNum;
+      if (expectedSliceSize > requiredMinimumBufferCapacity) {
+        return Pair.of(degree, requiredBufferNum);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Recursively compute number of required buffers in a top-down manner.
+   *
+   * @param totalNumNodes total number of descendent nodes
+   * @param degree        degree
+   *
+   * @return minimum number of buffers required for combining tree
+   */
+  private static int computeRequiredBufferNum(int totalNumNodes, int degree)
+  {
+    if (totalNumNodes > degree) {
+      final int numNodesPerChild = (totalNumNodes + degree - 1) / degree; // ceiling
+      int sum = 1; // count for the current node
+      for (int i = 0; i < degree; i++) {
+        // further compute for child nodes
+        sum += computeRequiredBufferNum(Math.min(numNodesPerChild, totalNumNodes - i * numNodesPerChild), degree);
+      }
+
+      return sum;
+    } else {
+      return 1;
+    }
+  }
+
+  /**
+   * Recursively build a combining tree in a top-down manner.
+   *
+   * @param sortedIterators    sorted iterators
+   * @param bufferSupplier     combining buffer supplier
+   * @param combiningFactories array of combining aggregator factories
+   * @param combineDegree      combining degree
+   *
+   * @return an iterator of the root of the combining tree
+   */
+  private Iterator<Entry<KeyType>> buildCombineTree(
```

Comment:
Better to return a pair of `Iterator` and `List<Future>`, then provide `List<Future>` "to fill"

Reply:
Changed.